File: src/com/train/algorithm/twoPointers/3SumClosest.cpp
//
// Created by 罗斯 on 31/8/2022.
//
#include <vector>

using namespace std;

class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        int n = nums.size();
        int d = INT_MAX;
        int res = target;
        sort(begin(nums), end(nums));

        for(int i = 0;i < n - 2;++i){
            int s = i + 1,t = n - 1;
            while (s < t){
                int sum = nums[i] + nums[s] + nums[t];
                if(sum == target){
                    return target;
                }
                int diff = abs(target - sum);
                if(diff < d){
                    d = diff;
                    res = sum;
                }
                if(sum > target) --t;
                else ++s;
            }
        }
        return res;
    }
};

File: src/com/train/algorithm/twoPointers/implementInPython/3SumClosest.py
class Solution(object):
    def threeSumClosest(self, nums, target):
        n = len(nums)
        if n == 3:
            return sum(nums)
        nums = sorted(nums)
        minSum, minDif = None, float('inf')

        for i in range(n - 2):
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            l, r = i + 1, n - 1

            while l < r:
                cSum = sum([nums[i], nums[l], nums[r]])
                cDif = abs(target - cSum)

                if cDif < minDif:
                    minDif = cDif
                    minSum = cSum

                    if minDif == 0:
                        return minSum

                if cSum < target:
                    l += 1
                    while l < r and nums[l] == nums[l - 1]:
                        l += 1
                else:
                    r -= 1
                    while l < r and nums[r] == nums[r + 1]:
                        r -= 1

        return minSum